src/neo-vm/ExecutionEngine.cs

```diff
@@ -1157,6 +1204,92 @@ private bool ExecuteInstruction()
             return true;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ExecuteTry(int catchOffset, int finallyOffset)
+        {
+            if (catchOffset == 0 && finallyOffset == 0) return false;
+
+            CurrentContext.TryStack ??= new Stack<TryContext>();
+            CurrentContext.TryStack.Push(new TryContext(CurrentContext, catchOffset, finallyOffset));
+            CurrentContext.MoveNext();
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal bool ExecuteEndTryCatch(TryState currentState)
+        {
+            if (CurrentContext.TryStack is null) return false;
+            if (!CurrentContext.TryStack.TryPeek(out TryContext currentTry))
+                return false;
+            if (currentTry.State != currentState) return false;
+
+            if (currentTry.HasFinally)
+            {
+                currentTry.State = TryState.Finally;
+                currentTry.EndTryCatch(CurrentContext.InstructionPointer + CurrentContext.CurrentInstruction.Size);
+                CurrentContext.InstructionPointer = currentTry.FinallyPointer;
+            }
+            else
+            {
+                CurrentContext.TryStack.TryPop(out _);
+                CurrentContext.MoveNext();
+            }
+            if (currentTry.State == TryState.Catch)
+            {
+                FaultState.Exception = null;
+            }
+            return true;
+        }
+
+        private bool HandleException()
+        {
+            foreach (var executionContext in InvocationStack)
+            {
+                if (executionContext.TryStack is null) continue;
+
+                while (executionContext.TryStack.TryPeek(out TryContext tryContext))
+                {
+                    if (tryContext.State == TryState.Finally || (tryContext.State == TryState.Catch && !tryContext.HasFinally))
+                    {
+                        executionContext.TryStack.Pop();
+                        continue;
+                    }
+
+                    ResumeContext(tryContext);
+                    if (tryContext.State == TryState.Try && tryContext.HasCatch)
+                    {
+                        tryContext.State = TryState.Catch;
+                        CurrentContext.InstructionPointer = tryContext.CatchPointer;
+                    }
+                    else
+                    {
+                        tryContext.State = TryState.Finally;
+                        CurrentContext.InstructionPointer = tryContext.FinallyPointer;
+                        FaultState.Rethrow = true;
+                    }
+
+                    State = VMState.NONE;
```

Comment:
The state should not be modified, right?

Reply:
You are right, done.